### ISSUE TITLE
Subsample encryption: Clarify whether frame_header_obu() in frame_obu() should be protected or unprotected

### ISSUE THREAD

**Issue by kqyang** [AOMediaCodec/av1-isobmff#88]
frame_obu() is basically frame_header_obu() + tile_group_obu(). 

[5.10. Frame OBU syntax](https://aomediacodec.github.io/av1-spec/av1-spec.pdf#page=71)
```
frame_obu( sz ) {
  startBitPos = get_position( )
  frame_header_obu( )
  byte_alignment( )
  endBitPos = get_position( )
  headerBytes = (endBitPos - startBitPos) / 8
  sz -= headerBytes
  tile_group_obu( sz )
}
```

https://aomediacodec.github.io/av1-isobmff/#subsample-encryption says that 

> ..., and Frame Header OBUs SHALL be unprotected.
>
> Tile Group OBUs and Frame OBUs SHALL be protected.

Does it mean that frame_obu() is protected including frame_header_obu() inside frame_obu()?

**Comment by cconcolato** [comment]
The spec also says:
> Within Tile Group OBUs or Frame OBUs, the following applies: [...] BytesOfProtectedData SHALL span all complete 16-byte blocks of the decode_tile structure (including any trailing bits) [...]

The intent was that only the decode_tile structure is allowed to be protected, but I agree this is not precise enough. It could say:
> BytesOfProtectedData SHALL not start before the first byte of the associated decode_tile structure.

Actually, a previous version said:
> Frame Header OBUs (including with a Frame OBU) SHALL be unencrypted. Tile Group OBUs, Frame OBUs (except the Frame Header OBU part) MAY be encrypted.

I could reinstate some text along those lines. Do you have a preference? Both?

**Comment by kqyang** [comment]
Thanks for the quick reply!

I prefer the first one. I think it also helps if the clause "Within Tile Group OBUs or Frame OBUs, the following applies" can be moved under "Tile Group OBUs and Frame OBUs SHALL be protected". 

Here is a proposal:

- all OBU Headers and associated obu_size fields SHALL be unprotected.

- Temporal Delimiter OBUs, Sequence Header OBUs, and Frame Header OBUs (including with a Frame OBU) SHALL be unprotected.

- Metadata OBUs MAY be protected.

- Tile Group OBUs and Frame OBUs SHALL be protected. Within Tile Group OBUs or Frame OBUs, the following applies:

  - A subsample SHALL be created for each tile.

  - BytesOfProtectedData SHALL start from the decode_tile structure with block (16 bytes) alignment. Block alignment may require that the start of BytesOfProtectedData is not at the first byte of decode_tile, but some number of bytes following that.

  - BytesOfProtectedData SHALL end on the last byte of the OBU for the last tile or on the last byte of decode_tile structures (including any trailing bits) for all other tiles. 

  - BytesOfProtectedData SHALL be a multiple of 16 bytes.

  - All other parts of Tile Group OBUs and Frame OBUs SHALL be unencrypted.

**Comment by kqyang** [comment]
Actually it does not seem to be necessary to call out the last tile in 

> BytesOfProtectedData SHALL end on the last byte of the OBU for the last tile or on the last byte of decode_tile structures (including any trailing bits) for all other tiles.

since the last tile also ends with decode_tile, so it should be sufficient to say:

- BytesOfProtectedData SHALL end on the last byte of decode_tile structures (including any trailing bits).

**Comment by cconcolato** [comment]
Your concern should be addressed in #90.

**Comment by porcelijn** [comment]
Wouldn't it make sense to simply prohibit Frame OBUs? If the av1-isobmff spec were to mandate separate OBUs for Frame Headers and Tile Groups that would greatly simplify expression of what needs to be encrypted at cost of adding a couple of bytes overhead.

What I'm suggesting is that encryption would apply _only_ to Tile Group OBUs and could simply state BytesOfProtectedData is (multiple of 16 bytes) following `obu_header()` and size field `leb128()`.

This would allow player side decryption process to iterate over OBUs, scanning for tile groups without having to worry about state (ie. `SeenFrameHeader`) or possibly executing a full `uncompressed_header()` parse _just_ to find the embedded Tile Group.

Similarly, the (just in time) packaging process would be able to do a quick & dumb scan of the elementary streams in the mp4 file to apply encryption. But the presence of Frame OBUs in the ES breaks this separation of concerns and incurs tight coupling with the AV1 format, which is error prone and not future proof.

As a compromise to mitigate some of the overhead concerns, we could also allow `OBU_FRAME`, but strictly limit it's use to those where `SeenFrameHeader == 1`. In this case the `OBU_FRAME` is effectively an alias for _empty_ `OBU_FRAME_HEADER` + `OBU_TILE_GROUP` and the rules for BytesOfProtectedData are identical earlier sketched Tile Group OBU rules.

**Comment by cconcolato** [comment]
> Wouldn't it make sense to simply prohibit Frame OBUs?

Anyone may do that in a specific profile, but in general we cannot.

> If the av1-isobmff spec were to mandate separate OBUs for Frame Headers and Tile Groups that would greatly simplify expression of what needs to be encrypted at cost of adding a couple of bytes overhead.

I agree that it would simplify the process of encrypting, but unfortunately, we cannot mandate that in the ISOBMFF spec. Applications wishing to simplify encryption (e.g. for live streaming systems) could do that.

> could simply state BytesOfProtectedData is (multiple of 16 bytes) following obu_header() and size field leb128()

It was discussed and decided to encrypt tiles, not tile groups. see #81 

> This would allow player side decryption process [...]

Players don't have to parse any OBU to do the decryption, they follow the subsample description. The complexity is only at the writer.

> But the presence of Frame OBUs in the ES breaks this separation

I agree it would have been helpful to be able to skip Frame Header OBUs within a Frame OBU but unfortunately, that's not the case. Also, the use of `SeenFrameHeader` in the parsing is unfortunate. Instead of using a Frame OBU when the Frame Header has been sent, one could simply send a Tile Group OBU. Unfortunately, this is allowed by the AV1 video spec.

Summary, I understand your concern but we cannot limit the spec. It has to support all streams that can possibly be encoded (except those Tile List OBU).

**Comment by rbouqueau** [comment]
> while transformation from OBU_FRAME_HEADER + OBU_TILE_GROUP to OBU_FRAME is trivial, the reverse requires full stream parsing. This complexity could be avoided by simply adding OBU_FRAME to above list of excluded OBUs. 

Agreed.

**Comment by cconcolato** [comment]
> I think we could and there is already a precedent for excluding certain OBU types

Not that this says "SHOULD NOT", it is not forbidding them.

> For any OBU_FRAME, there is an equivalent OBU_FRAME_HEADER + OBU_TILE_GROUP sequence. [...]

I don't disagree on the complexity and overhead aspects. I'm just saying that we cannot forbid OBU_FRAME completely. I am not opposed to a recommendation to not use OBU_FRAME to reduce complexity. I will propose that during today's AOM call and we will decide.

**Comment by cconcolato** [comment]
> Can it do so only through senc information (or saiz/saio)?

Yes, through sample auxiliary information.



> Would this be signaled in a SampleEntry?

Do you mean, if only metadata were encrypted, what would you signal? I think as long as a single OBU is encrypted, you have to signal that the stream is encrypted. Maybe you will miss the opportunity for some decoders to decode everything but the metadata because they cannot decrypt but I don't think that's a frequent use case.

**Comment by kqyang** [comment]
Re https://github.com/AOMediaCodec/av1-isobmff/issues/88#issuecomment-417634147,

> It is certainly possible to use subs (or saiz, saio, senc) so that the client have to, but Common Encryption does not dictate that these boxes be present, and neither does our spec. Indeed, Apple choose to ignore these boxes in their adoption of CBCS and simply follow the AVC (and HEVC) definitions for finding the subsample ranges.

Common Encryption spec does require the presence of Sample Auxiliary Information (explicit for cbcs, implicit for cenc):

> 10.4 ‘cbcs’ AES-CBC subsample pattern encryption scheme
...
Unless it would be empty, as noted in 7.1, the Sample Auxiliary Information SHALL be present and
SHALL identify protected ranges as Subsamples.

A decryptor can decrypt by parsing OBUs or simply use subsample information carried in Sample Auxiliary Information.

> Indeed, Apple choose to ignore these boxes in their adoption of CBCS and simply follow the AVC (and HEVC) definitions for finding the subsample ranges.

I am not surprised that if they do it this way as that is what they do with their legacy SAMPLE-AES decryption, which does not have subsample information. 

Re https://github.com/AOMediaCodec/av1-isobmff/issues/88#issuecomment-417634147,

> You are correct, thanks for pointing that out. What I meant to express was that by excluding frame OBUs at least it becomes trivial to find the start of the tile group whose tiles need to be encrypted, because anything except OBU_TILE_GROUP can be ignored.

Unfortunately, there are still a lot of complexities, since you'll still need to parse `frame_header_obu()` to extract tile info, which is needed to do per tile subsample encryption. It does simplify things a bit so I support the recommendation to not use OBU_FRAME.

**Comment by cconcolato** [comment]
The AOM Codec WG discussed this issue and in particular the last comment about adding a NOTE but it was decided not to do it for v1.0.0.

**Comment by porcelijn** [comment]
Re https://github.com/AOMediaCodec/av1-isobmff/issues/88#issuecomment-417745438 @kqyang,
> Common Encryption spec does require the presence of Sample Auxiliary Information (explicit for cbcs, implicit for cenc):
>
> >   10.4 ‘cbcs’ AES-CBC subsample pattern encryption scheme
> >   ...
> >   Unless it would be empty, as noted in 7.1, the Sample Auxiliary Information SHALL be present and SHALL identify protected ranges as Subsamples.

My bad, I only read as far as 7.1:

> ... Both Initialization Vectors and Subsample encryption information MAY be provided as Sample Auxiliary Information [...]

Nevertheless, later on in paragraph 10.4 you cited, it mentions:

> ... It is therefore possible to decrypt a track using either (a) this algorithm, ignoring the Sample Auxiliary Information​ or​ (b)​ the​ Sample​ Auxiliary Information,​​ ignoring​​ this​​ algorithm.

What I'm trying to point out, then, is that option (a) is complicated for `OBU_FRAME` and essentially *broken* by the _ambiguous_ description around:

> Metadata OBUs MAY be protected.

When following the algorithm approach, how do I prevent mistakenly decrypting metadata OBUs that were never encrypted? This may seem minor, but I believe we have a problem if at all times the decoder MUST rely on Sample Auxiliary Information, rather than our algorithm; our definitions: https://aomediacodec.github.io/av1-isobmff/#subsample-encryption Especially since we're about to release v1.0.0.

At minimum, the spec should warn that we intend to violate this constraint of Common Encryption and _only_ allow decryption the lines of 10.4.1(b).

But IMO, we should also put more effort into shaving off the rough edges https://aomediacodec.github.io/av1-isobmff/#CommonEncryption . I suppose there are both use cases where metadata should be in the clear and others where it's protected but the spec should define how these are discriminated or signaled to _both_ packager and decoder. Similarly, I reiterate my call add `OBU_FRAME` to the SHOULD NOT use list.

**Comment by porcelijn** [comment]
@cconcolato, while the https://aomediacodec.github.io/av1-isobmff link hasn't been updated, I'm assuming this issue's status is now Closed as a "won't fix in v1.0.0", but I'm hoping to continue the conversation. Will you reopen this issue, should I create a new one or am I just being impatient (I noticed yesterday was a US holiday) :confused:

**Comment by cconcolato** [comment]
>  I believe we have a problem if at all times the decoder MUST rely on Sample Auxiliary Information, rather than our algorithm; our definitions

I don't follow you. I don't see where the algorithm and CENC differ.

> At minimum, the spec should warn that we intend to violate this constraint of Common Encryption and only allow decryption along the lines of 10.4.1(b).

There is no violation of Common Encryption in the AV1-ISOBMFF spec. Can you clarify what you mean?

> @cconcolato, while the https://aomediacodec.github.io/av1-isobmff link hasn't been updated, 

We are about to publish v1.0.0, see https://github.com/AOMediaCodec/av1-isobmff/pull/99

> I'm assuming this issue's status is now Closed as a "won't fix in v1.0.0", 

Yes

> but I'm hoping to continue the conversation. Will you reopen this issue, should I create a new one 

I will not reopen this issue as everything has been said (on SAI, Metadata and Frame OBUs) and I don't see a problem. It's up to you to create a new one or continue here.

> or am I just being impatient (I noticed yesterday was a US holiday) 😕

Maybe ;)

**Comment by porcelijn** [comment]
> There is no violation of Common Encryption in the AV1-ISOBMFF spec. Can you clarify what you mean?

From https://github.com/AOMediaCodec/av1-isobmff/issues/88#issuecomment-418050388:

> [...] in paragraph 10.4 you cited, it mentions:
>
> >    ... It is therefore possible to decrypt a track using either (a) this algorithm, ignoring the Sample Auxiliary Information​ or​ (b)​ the​ Sample​ Auxiliary Information,​​ ignoring​​ this​​ algorithm.

Unless I misinterpret this, Common Encryption says that a client could opt to (a) _ignore_ Sample Auxiliary Information and instead decrypt using _only_ the algorithm (that is: the rules in our spec https://aomediacodec.github.io/av1-isobmff/#subsample-encryption). But this approach is impossibe if encoder has the freedom to choose to encrypt metadata and inform the client about this choice only implicitly through SAI.

Your line of reasoning seems to be that SAI is mandatory anyway, so why would the client _not_ use it? And while that arguably makes sense, our spec should mention that client must follow:

> [...] (b)​ the​ Sample​ Auxiliary Information,​​ ignoring​​ this​​ algorithm.

if it wants to avoid problems in current spec (and/or possible incompatibilities in future versions of av1-isobmff-cenc).

I think that SAIs in iso 23001-7 were intended as a convenience supplement, not an integral part of the protection itself.

> I will not reopen this issue as everything has been said (on SAI, Metadata and Frame OBUs) and I don't see a problem.

Personally I think that's unfortunate. But since there's already consensus and v1.0.0 is (about to be?) released let's indeed close the issue and work around the problems as they (may) show up.
